*** ../binutils-2.24-2013-10-25.orig/binutils/nm.c	2013-10-25 10:39:21.701010088 +0100
--- binutils/nm.c	2013-10-25 10:41:32.789013721 +0100
*************** display_rel_file (bfd *abfd, bfd *archiv
*** 1010,1016 ****
  
    symcount = bfd_read_minisymbols (abfd, dynamic, &minisyms, &size);
    if (symcount < 0)
!     bfd_fatal (bfd_get_filename (abfd));
  
    if (symcount == 0)
      {
--- 1010,1024 ----
  
    symcount = bfd_read_minisymbols (abfd, dynamic, &minisyms, &size);
    if (symcount < 0)
!     {
!       if (dynamic && bfd_get_error () == bfd_error_no_symbols)
! 	{
! 	  non_fatal (_("%s: no symbols"), bfd_get_filename (abfd));
! 	  return;
! 	}
! 
!       bfd_fatal (bfd_get_filename (abfd));
!     }
  
    if (symcount == 0)
      {
