*** ../binutils-2.23.51.0.8.orig/gas/config/tc-arm.c	2013-01-21 10:21:58.492233640 +0000
--- gas/config/tc-arm.c	2013-01-21 10:23:21.640318334 +0000
*************** const char FLT_CHARS[] = "rRsSfFdDxXeEpP
*** 885,890 ****
--- 885,893 ----
  static inline int
  skip_past_char (char ** str, char c)
  {
+   /* PR gas/14987: Allow for whitespace before the expected character.  */
+   skip_whitespace (*str);
+   
    if (**str == c)
      {
        (*str)++;
*************** parse_address_main (char **str, int i, i
*** 5168,5173 ****
--- 5171,5179 ----
        return PARSE_OPERAND_SUCCESS;
      }
  
+   /* PR gas/14887: Allow for whitespace after the opening bracket.  */
+   skip_whitespace (p);
+ 
    if ((reg = arm_reg_parse (&p, REG_TYPE_RN)) == FAIL)
      {
        inst.error = _(reg_expected_msgs[REG_TYPE_RN]);
