It reverts the H.J. Lu's binutils custom patch.

Index: ld/NEWS
===================================================================
--- ld/NEWS.orig	2009-02-04 19:21:50.000000000 +0100
+++ ld/NEWS	2009-02-23 16:18:35.000000000 +0100
@@ -26,9 +26,6 @@
   For the switch --enable-runtime-pseudo-reloc it uses for 32-bit
   runtime pseudo relocation version one, for 64-bit the version two.
 
-* ELF: Support environment variables, LD_SYMBOLIC for -Bsymbolic and
-  LD_SYMBOLIC_FUNCTIONS for -Bsymbolic-functions.
-
 Changes in 2.19:
 
 * Linker scripts support a new INSERT command that makes it easier to
Index: ld/ld.texinfo
===================================================================
--- ld/ld.texinfo.orig	2009-02-04 19:21:50.000000000 +0100
+++ ld/ld.texinfo	2009-02-23 16:18:06.000000000 +0100
@@ -1163,21 +1163,14 @@ When creating a shared library, bind ref
 definition within the shared library, if any.  Normally, it is possible
 for a program linked against a shared library to override the definition
 within the shared library.  This option is only meaningful on ELF
-platforms which support shared libraries.  If @option{-Bsymbolic} is not
-used when linking a shared library, the linker will also turn on this
-option if the environment variable @code{LD_SYMBOLIC} is set.
+platforms which support shared libraries.
 
 @kindex -Bsymbolic-functions
 @item -Bsymbolic-functions
 When creating a shared library, bind references to global function
 symbols to the definition within the shared library, if any.
 This option is only meaningful on ELF platforms which support shared
-libraries.  If @option{-Bsymbolic-functions} is not used when linking a
-shared library, the linker will also turn on this option if the
-environment variable @code{LD_SYMBOLIC_FUNCTIONS} is set.  When
-both environment variables @code{LD_SYMBOLIC} and
-@code{LD_SYMBOLIC_FUNCTIONS} are set, @code{LD_SYMBOLIC} will take
-precedent.
+libraries.
 
 @kindex --dynamic-list=@var{dynamic-list-file}
 @item --dynamic-list=@var{dynamic-list-file}
Index: ld/ldmain.c
===================================================================
--- ld/ldmain.c.orig	2009-02-04 19:21:50.000000000 +0100
+++ ld/ldmain.c	2009-02-23 16:19:01.000000000 +0100
@@ -256,11 +256,6 @@ main (int argc, char **argv)
   command_line.warn_search_mismatch = TRUE;
   command_line.check_section_addresses = -1;
 
-  if (getenv ("LD_SYMBOLIC") != NULL)
-    command_line.symbolic = symbolic;
-  else if (getenv ("LD_SYMBOLIC_FUNCTIONS") != NULL)
-    command_line.symbolic = symbolic_functions;
-
   /* We initialize DEMANGLING based on the environment variable
      COLLECT_NO_DEMANGLE.  The gcc collect2 program will demangle the
      output of the linker, unless COLLECT_NO_DEMANGLE is set in the
