*** ../orig.binutils-2.23.51.0.1/binutils/elfcomm.c	2013-09-09 13:01:22.010343521 +0100
--- binutils/elfcomm.c	2013-09-09 13:01:59.389344557 +0100
*************** error (const char *message, ...)
*** 36,41 ****
--- 36,42 ----
  {
    va_list args;
  
+   fflush (stdout);
    va_start (args, message);
    fprintf (stderr, _("%s: Error: "), program_name);
    vfprintf (stderr, message, args);
*************** warn (const char *message, ...)
*** 47,52 ****
--- 48,54 ----
  {
    va_list args;
  
+   fflush (stdout);
    va_start (args, message);
    fprintf (stderr, _("%s: Warning: "), program_name);
    vfprintf (stderr, message, args);
