*** ../binutils-2.20.51.0.10.orig/ld/ldlex.l	2010-08-20 15:28:06.000000000 +0100
--- ld/ldlex.l	2010-08-20 15:35:27.000000000 +0100
*************** CMDFILENAMECHAR1  [_a-zA-Z0-9\/\.\\_\+\$
*** 99,105 ****
  FILENAMECHAR1	[_a-zA-Z\/\.\\\$\_\~]
  SYMBOLCHARN     [_a-zA-Z\/\.\\\$\_\~0-9]
  FILENAMECHAR	[_a-zA-Z0-9\/\.\-\_\+\=\$\:\[\]\\\,\~]
! WILDCHAR	[_a-zA-Z0-9\/\.\-\_\+\=\$\:\[\]\\\,\~\?\*]
  WHITE		[ \t\n\r]+
  
  NOCFILENAMECHAR	[_a-zA-Z0-9\/\.\-\_\+\$\:\[\]\\\~]
--- 99,105 ----
  FILENAMECHAR1	[_a-zA-Z\/\.\\\$\_\~]
  SYMBOLCHARN     [_a-zA-Z\/\.\\\$\_\~0-9]
  FILENAMECHAR	[_a-zA-Z0-9\/\.\-\_\+\=\$\:\[\]\\\,\~]
! WILDCHAR	[_a-zA-Z0-9\/\.\-\_\+\=\$\:\[\]\\\,\~\?\*\^\!]
  WHITE		[ \t\n\r]+
  
  NOCFILENAMECHAR	[_a-zA-Z0-9\/\.\-\_\+\$\:\[\]\\\~]
